src/apps/MASS3DEA.hpp

```diff
@@ -0,0 +1,133 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2017-23, Lawrence Livermore National Security, LLC
+// and RAJA Performance Suite project contributors.
+// See the RAJAPerf/LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+///
+/// Action of 3D mass matrix via partial assembly
+///
+/// Based on MFEM's/CEED algorithms.
+/// Reference implementation
+/// https://github.com/mfem/mfem/blob/master/fem/bilininteg_mass_ea.cpp#L142
+///
+/// for (int e = 0; e < NE; ++e) {
+///
+///
+/// } // element loop
+///
+
+#ifndef RAJAPerf_Apps_MASS3DEA_HPP
+#define RAJAPerf_Apps_MASS3DEA_HPP
+
+#define MASS3DEA_DATA_SETUP                                                    \
+  Real_ptr B = m_B;                                                            \
+  Real_ptr D = m_D;                                                            \
+  Real_ptr M = m_M;                                                            \
+  Index_type NE = m_NE;
+
+#include "common/KernelBase.hpp"
+#include "FEM_MACROS.hpp"
+
+#include "RAJA/RAJA.hpp"
+
+// Number of Dofs/Qpts in 1D
+#define MEA_D1D 4
+#define MEA_Q1D 5
+#define B_MEA_(x, y) B[x + MEA_Q1D * y]
+#define M_(i1, i2, i3, j1, j2, j3, e)                                          \
+  M[i1 + MEA_D1D * (i2 + MEA_D1D * (i3 + MEA_D1D * (j1 + MEA_D1D * (j2 + MEA_D1D * (j3 + MEA_D1D * e)))))]
+
+#define D_MEA_(qx, qy, qz, e)                                                      \
+  D[qx + MEA_Q1D * qy + MEA_Q1D * MEA_Q1D * qz +                               \
+    MEA_Q1D * MEA_Q1D * MEA_Q1D * e]
+
+#define MASS3DEA_0_CPU double s_B[MEA_Q1D][MEA_D1D];
+
+#define MASS3DEA_0_GPU __shared__ double s_B[MEA_Q1D][MEA_D1D];
+
+#define MASS3DEA_0 RAJA_TEAM_SHARED double s_B[MEA_Q1D][MEA_D1D];
+
+#define MASS3DEA_1 s_B[q][d] = B_MEA_(q, d);
+
+#define MASS3DEA_2_CPU                                                         \
```

Comment:
Should these three macros be changed to avoid redundancy? It looks like the only difference is the qualifier indicating whether to use shared memory and which type of shared memory.  Maybe just have a macro for the memory type?